### PR TITLE
mimic: rgw_file:  fix invalidation of top-level directories

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/40215

### CODE REVIEW

**Comment by smithfarm** [comment]
Build failure is a huge wall of text starting with:

```
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_os_lib.cc:8:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_file.h: In member function 'rgw::RGWFileHandle* rgw::RGWLibFS::lookup_handle(rgw_fh_hk)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_file.h:1200:44: error: no match for 'operator<<' (operand types are 'std::basic_ostream<char>' and 'rgw::fh_key')
    << __func__ << " handle lookup failed " << fhk
In file included from /usr/include/c++/8/iterator:64,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/utility/string_ref.hpp:26,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rest.h:9,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_os_lib.cc:4:
/usr/include/c++/8/ostream:108:7: note: candidate: 'std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::basic_ostream<_CharT, _Traits>::__ostream_type& (*)(std::basic_ostream<_CharT, _Traits>::__ostream_type&)) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]'
       operator<<(__ostream_type& (*__pf)(__ostream_type&))
       ^~~~~~~~
```

**Comment by smithfarm** [comment]
That build failure seems really weird to me. Re-running the test.

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
@mattbenjamin The "wall-of-text" build failure is reproducible, and I'm afraid I haven't the faintest notion what is causing it. Could you take a look?

**Comment by smithfarm** [comment]
@chardan This cherry-pick applied cleanly, but then the build emitted a desperate wall of text and failed. I looked at it from several different directions but came up empty-handed. Any idea what the problem is?

The "wall of text" (build errors) begins with https://github.com/ceph/ceph/pull/29276#issuecomment-514708056 and continues in the same vein. Somehow it think the operator `<<` is not defined in the given context, but that makes no sense?

**Comment by smithfarm** [comment]
I preserved the "wall-of-text" build errors in a gist so they'll still be accessible after Jenkins deletes the build log: https://gist.github.com/smithfarm/7e34a42dd217690278f069725f6c0582

**Comment by chardan** [comment]
Hi Nathan, I'll try to help out more later today, but the main thing the compiler's complaining about is that there's no operator<<() defined for the type "rgw::fh_key". The rest of the message is basically GCC explaining everything it tried to match before it gave up. Hope that's useful!

**Comment by smithfarm** [comment]
Thanks, @chardan, that's helpful - I understand the issue much better now!

**Comment by smithfarm** [comment]
jenkins test docs

File: src/rgw/rgw_file.h [hunk 1]
```diff
@@ -1191,10 +1194,13 @@ namespace rgw {
 			    RGWFileHandle::FHCache::FLAG_LOCK);
       /* LATCHED */
       if (! fh) {
+	if (unlikely(fhk == root_fh.fh.fh_hk)) {
```
Comment (yangruifeng) [inline src/rgw/rgw_file.h]: what's different from the following code（line1127）

Comment (smithfarm) [reply src/rgw/rgw_file.h]: @yangruifeng I don't entirely understand your question, but this is a backport so the similarity with the code in master is intentional.

Comment (yangruifeng) [reply src/rgw/rgw_file.h]: @smithfarm it's ok,  I just think it's a little redundant

**Comment by yuriw** [comment]
[wip-yuri2-testing-2019-10-04-0015-mimic](https://trello.com/c/xfBj13mW/707-wip-yuri2-testing-2019-10-04-0015-mimic)

**Pull request merged**